src/cart/crt_hg.c

```diff
@@ -746,6 +746,8 @@ crt_hg_class_init(int provider, int idx, bool primary, hg_class_t **ret_hg_class
 	if (rc != 0)
 		D_GOTO(out, rc);
 
+	init_info.na_init_info.auth_key = getenv("CRT_AUTH_KEY");
```

Comment:
We should check for the provider here and do this only for provider == CRT_PROV_OFI_CXI.

Also, as a rule we read out envariables in data_init() call and store values in crt_gdata struct.

New envariables should be documented in src/cart/README.env
As a last thing, envariables should also be added to dump_envariables() function to help with the debug later on.

Reply:
> We should check for the provider here and do this only for provider == CRT_PROV_OFI_CXI.
> 
Auth keys are used by more than just CXI though, no? Looks like maybe also the GNI provider. In any case, it seems like it would be forward compatible to just set it and let mercury decide what to do with it (if anything).

> Also, as a rule we read out envariables in data_init() call and store values in crt_gdata struct.
> 
So are you saying that you would like to see a new `char *auth_key` member of that struct?

> New envariables should be documented in src/cart/README.env 
>
OK, I can add that.

> As a last thing, envariables should also be added to dump_envariables() function to help with the debug later on.
>
I can do that, but I wonder if there are security considerations there.

Reply:
We don't use auth keys for anything else today. My concern is that if this gets set in env/scripts and starts causing issues with TCP when people forget to unset it during testing. We also don't support GNI, nor planning to. @soumagne what is the current mercury behavior for non cxi providers when auth key is set? 

Yes something like char auth_key[MAX_KEY_SIZE], so that we limit the input to something reasonable like max=255. Otherwise we might get flagged by SDL/codescans. 

Good point on security considerations, perhaps then lets skip that part for now.

Reply:
There is no need for auth keys for tcp so no need to worry about it. I think it's harmless to set it like that since even for tcp we already have a check for it internally and depending on the provider (gni, CXI) the auth key has a different format.

Reply:
> Another thing to think about here is whether this is sufficient for a multi-provider case. Do we need 2 different auth keys in case of CXI primary, CXI secondary? Primary and secondary would be on different networks.

My understanding is that there will be a single VNI for DAOS traffic. I suppose that we could try to future-proof it by making it support multiple keys from the beginning, but I'm not sure what that looks like at the mercury level. @soumagne: Any thoughts on this?

Reply:
Each cart context / mercury class will take a single key. But yes if in the future we have multiple CXI  domains (primary + secondary) as Alex introduced I could see a need for having multiple auth keys (though still only one would be used per context / class).

Reply:
I spent some time trying to add the requested support for multiple auth keys, but it quickly gets complicated because we need to deal with multiple strings. All of that went OK, but then I got to `crt_hg_class_init()` and realized that I needed to copy the auth key string so that it could be safely used in the `init_info` struct, but I didn't see a good way to free that string later on (there's no `crt_hg_class_fini()`) without adding more infrastructure.

As all of this stuff is well outside my area of expertise, I would prefer to land my simple solution here to make sure that we are not blocked on Aurora testing, and leave the correct solution to the networking experts. OK?

Reply:
Actually, after re-reading the code, I guess it's safe to free the auth key string after `HG_init_opt()` is finished, as the mercury stuff just does a `sscanf()` to parse the key into a uint32/uint16 pair. I was worried that the string was used elsewhere. I'll just finish my patch.

Reply:
I am a bit confused here -- The handling of multiple auth keys should be similar to handling of multiple provider, interface, domain and port strings. Those are done in crt_init_opt():
So something like:
`auth_env = get_env("CRT_AUTH_KEY");

rc = __split_arg(auth_env, &key0, &key1);

rc = crt_na_config_init(true, primary_provider, iface0, domain0, port0, key0);`

And similarly for initing secondary providers (we only support 1 for now) for key1.

Reply:
Added the comment before the latest patch:) will re-read